File: C++/Array&String/345. Reverse Vowels of a String.cpp
// 345. Reverse Vowels of a String
// Difficulty : Easy
// Submission Link: https://leetcode.com/problems/reverse-vowels-of-a-string/submissions/1017549314/

class Solution {
public:
    string reverseVowels(string s) {
        int left = 0;
        int right = s.length() - 1;
        unordered_set<char> vowels = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'};

        while (left < right) {
            while (left < right && vowels.find(s[left]) == vowels.end()) {
                left++;
            }

            while (left < right && vowels.find(s[right]) == vowels.end()) {
                right--;
            }

            if (left < right) {
                swap(s[left], s[right]);
                left++;
                right--;
            }
        }

        return s;
    }
};

File: Python/Array&String/345. Reverse Vowels of a String.py
""" 345. Reverse Vowels of a String
Difficulty: Easy
Submission Link: https://leetcode.com/problems/reverse-vowels-of-a-string/submissions/1017551627
 """

class Solution(object):
    def reverseVowels(self, s):
        vowels = "aeiouAEIOU"
        s = list(s)
        left, right = 0, len(s) - 1
        
        while left < right:
            while left < right and s[left] not in vowels:
                left += 1
            while left < right and s[right] not in vowels:
                right -= 1
            
            s[left], s[right] = s[right], s[left]
            left += 1
            right -= 1
        
        return ''.join(s)